File: simple-factory/SimpleFactory.py
from enum import Enum
# python 3.8 and above
from typing import Protocol, Dict

class Material(Enum):
    WOOD = 1
    STONE = 2
    IRON = 3

# protocol is a new feature in python 3.8, it is similar to interface in java
# Note that it is only for type checking, it does not enforce the implementation
# https://docs.python.org/3/library/typing.html#typing.Protocol

class Axe(Protocol):
    def getMaterial(self):
        pass

class WoodAxe(Axe):
    def getMaterial(self):
        return "wood"

class StoneAxe(Axe):
    def getMaterial(self):
        return "stone"

class IronAxe(Axe):
    def getMaterial(self):
        return "iron"

class CraftingFactory:
    def createAxe(self, material: Material):
        craftMap: Dict[Material, Axe] = {
            Material.WOOD: WoodAxe,
            Material.STONE: StoneAxe,
            Material.IRON: IronAxe
        }
        if(material not in craftMap):
            raise Exception("Invalid material")
        return craftMap[material]()

def testCraftingFactory():
    factory = CraftingFactory()
    woodAxe = factory.createAxe(Material.WOOD)
    stoneAxe = factory.createAxe(Material.STONE)
    ironAxe = factory.createAxe(Material.IRON)

    print("WoodAxe is made of " + woodAxe.getMaterial())
    print("StoneAxe is made of " + stoneAxe.getMaterial())
    print("IronAxe is made of " + ironAxe.getMaterial())

if __name__ == "__main__":
    testCraftingFactory()

File: simple-factory/SimpleFactory.java

public class SimpleFactory {

    /* Axe Class */
    private interface Axe {
        String getMaterial();
    }

    private enum Material {
        WOOD,
        STONE,
        IRON
    }

    // Wooden Axe
    private class WoodenAxe implements Axe {
        @Override
        public String getMaterial() {
            return "wood";
        }
    }

    // Stone Axe
    private class StoneAxe implements Axe {
        @Override
        public String getMaterial() {
            return "stone";
        }
    }

    // Iron Axe
    private class IronAxe implements Axe {
        @Override
        public String getMaterial() {
            return "iron";
        }
    }

    /* Simple factory*/
    public class CraftingFactory {

        public Axe createAxe(Material material) {
            Axe axe;
            switch (material) {
                case WOOD:
                    axe = new WoodenAxe();
                    break;
                case STONE:
                    axe = new StoneAxe();
                    break;
                case IRON:
                    axe = new IronAxe();
                    break;
                default:
                    throw new IllegalArgumentException("Invalid material");
            }
            return axe;
        }
    }

    public void testCraftingFactory() {
        CraftingFactory  craftinTable = new CraftingFactory();
        
        var woodAxe = craftinTable.createAxe(Material.WOOD);
        var stoneAxe = craftinTable.createAxe(Material.STONE);
        var ironAxe = craftinTable.createAxe(Material.IRON);

        System.out.println("WoodAxe is made of " + woodAxe.getMaterial());
        System.out.println("StoneAxe is made of " + stoneAxe.getMaterial());
        System.out.println("IronAxe is made of " + ironAxe.getMaterial());
    }

    public static void main(String[] args) {
        SimpleFactory simpleFactory = new SimpleFactory();
        simpleFactory.testCraftingFactory();
    }
}